tasks/src/lib.rs

```diff
@@ -26,11 +27,11 @@ use thiserror::Error;
 /// and move out uninteresting responses (returning `Some(...)`).
 ///
 /// If no active task is interested in a given response, we call this response "unsolicited".
-pub trait Task: 'static {
+pub trait Task: Send + 'static {
```

Comment:
Why is `Send` necessary?

Reply:
I test my changes directly from `email-lib`, and somehow it was complaining about `Send` missing. Since my use case is really basic (I believe it should be the case of most users), I thought it would not hurt to add the `Send` restriction.

Reply:
> I test my changes directly from `email-lib`, and somehow it was complaining about `Send` missing.

I think you can write `T: Task + Send` in your lib in places where it's necessary.

> Since my use case is really basic (I believe it would be the case of most users), I thought it would not hurt to add the `Send` restriction since `this trait is automatically implemented when the compiler determines it’s appropriate` [[sources](https://doc.rust-lang.org/std/marker/trait.Send.html)].

Yes, all current implementors of `Task` are `Send`. But adding this bound

- implies that `imap-tasks` somehow relies on this bound, which is not true and
- prevents us from adding implementors in the future that are not `Send` (e.g. because they borrow data instead of owning them).

If there is still a good reason to add this bound, than I'm okay with it. But if you can solve your problem by writing `T: Task + Send` in `email-lib`, then I think we should prefer that. If you point me to your problem in `email-lib` then maybe I can better understand the nature of the problem.

Reply:
I dug a bit more, and the issue comes from places where `Box<dyn Any>` is used, mostly `TaskAny::process_tagged` and `TaskMap::tasks`. In `email-code`, the backend struct that holds a scheduler (soon a client) have the `Send` restriction by nature, which "imposes" this restriction down. From `email-lib` I have no control over this behaviour. Here the minimum changes I needed to do in order to make the compiler happy: https://github.com/duesee/imap-flow/pull/174/commits/0da136d34d08adc455dbae472eecc6d49b796c69.

I don't see any happy path:

- Without `Send` restriction in `imap-tasks`, users working with `Send` restriction cannot use the Scheduler
- With `Send` restriction in `imap-tasks`, users working without `Send` restriction (`Rc`) cannot use the Scheduler
- Users cannot change this behaviour from their side

Any idea @duesee?

Reply:
So your problem is that `Scheduler` is not `Send`. I think that's a good reason to add the bound to `Task`. Then we should also add an [assertion](https://docs.rs/static_assertions/latest/static_assertions/macro.assert_impl_all.html) to ensure that `Scheduler` is `Send` and we don't break it accidentally.